File: tests/python/test_allreduce.py
"""AllReduce correctness tests for Unified-CCL Python API."""

import torch
import pytest


def test_allreduce_fp16():
    """Test AllReduce with float16 tensors."""
    try:
        import uccl
    except ImportError:
        pytest.skip("uccl module not available")

    comm = uccl.Communicator(rank=0, world_size=1)
    t = torch.ones(1024, dtype=torch.float16, device="xpu")
    result = comm.allreduce(t)

    # Single rank: result should equal input
    assert torch.allclose(result, torch.ones_like(result)), \
        "AllReduce fp16 failed: single rank result != input"
    comm.destroy()


def test_allreduce_bf16():
    """Test AllReduce with bfloat16 tensors."""
    try:
        import uccl
    except ImportError:
        pytest.skip("uccl module not available")

    comm = uccl.Communicator(rank=0, world_size=1)
    t = torch.ones(1024, dtype=torch.bfloat16, device="xpu")
    result = comm.allreduce(t)

    assert torch.allclose(result, torch.ones_like(result)), \
        "AllReduce bf16 failed: single rank result != input"
    comm.destroy()


def test_allreduce_various_sizes():
    """Test AllReduce with various tensor sizes."""
    try:
        import uccl
    except ImportError:
        pytest.skip("uccl module not available")

    comm = uccl.Communicator(rank=0, world_size=1)
    sizes = [1, 16, 256, 1024, 4096, 65536, 262144, 1048576]

    for size in sizes:
        t = torch.ones(size, dtype=torch.float16, device="xpu")
        result = comm.allreduce(t)
        assert torch.allclose(result, torch.ones_like(result)), \
            f"AllReduce failed for size={size}"

    comm.destroy()


def test_allreduce_inplace():
    """Test that AllReduce modifies tensor in-place."""
    try:
        import uccl
    except ImportError:
        pytest.skip("uccl module not available")

    comm = uccl.Communicator(rank=0, world_size=1)
    t = torch.ones(1024, dtype=torch.float16, device="xpu") * 42.0
    data_ptr_before = t.data_ptr()
    result = comm.allreduce(t)

    # Should be the same tensor (in-place)
    assert result.data_ptr() == data_ptr_before, \
        "AllReduce should modify tensor in-place"
    comm.destroy()


def test_communicator_lifecycle():
    """Test communicator creation and destruction."""
    try:
        import uccl
    except ImportError:
        pytest.skip("uccl module not available")

    comm = uccl.Communicator(rank=0, world_size=1)
    assert comm.rank == 0
    assert comm.world_size == 1
    assert repr(comm).startswith("Communicator(")

    comm.destroy()
    # Should be safe to call destroy again
    comm.destroy()


def test_invalid_dtype():
    """Test that unsupported dtype raises error."""
    try:
        import uccl
    except ImportError:
        pytest.skip("uccl module not available")

    comm = uccl.Communicator(rank=0, world_size=1)
    t = torch.ones(1024, dtype=torch.float32, device="xpu")

    with pytest.raises(AssertionError):
        comm.allreduce(t)

    comm.destroy()


def test_invalid_device():
    """Test that CPU tensor raises error."""
    try:
        import uccl
    except ImportError:
        pytest.skip("uccl module not available")

    comm = uccl.Communicator(rank=0, world_size=1)
    t = torch.ones(1024, dtype=torch.float16)  # CPU tensor

    with pytest.raises(AssertionError):
        comm.allreduce(t)

    comm.destroy()


def test_invalid_rank():
    """Test that invalid rank raises error."""
    try:
        import uccl
    except ImportError:
        pytest.skip("uccl module not available")

    with pytest.raises(ValueError):
        uccl.Communicator(rank=-1, world_size=1)

    with pytest.raises(ValueError):
        uccl.Communicator(rank=2, world_size=2)


def test_allreduce_multi_gpu():
    """Multi-process test using torch.multiprocessing.

    To be run with:
        mpirun -n <nranks> python -m pytest test_allreduce.py::test_allreduce_multi_gpu
    """
    # This test requires multi-process setup
    # Skipped in single-process test runner
    pytest.skip("Multi-GPU test requires mpirun")


if __name__ == "__main__":
    pytest.main([__file__, "-v"])

File: tests/cpp/test_allreduce.cc
#include <sycl/sycl.hpp>
#include <mpi.h>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <cmath>
#include <vector>

#include "uccl.h"

/* AllReduce correctness and performance test.
 * Mirrors nccl-tests all_reduce_perf.
 *
 * Usage:
 *   mpirun -n <nranks> ./test_allreduce
 *
 * Tests:
 * 1. Single rank: verify identity (result == input)
 * 2. Multi rank: verify sum across ranks
 * 3. Various message sizes for bandwidth measurement
 */

static bool checkResult(const sycl::half* data, size_t count,
                        float expected, float tolerance) {
    for (size_t i = 0; i < count; i++) {
        float val = static_cast<float>(data[i]);
        if (std::fabs(val - expected) > tolerance) {
            std::fprintf(stderr,
                "MISMATCH at index %zu: got %f, expected %f\n",
                i, val, expected);
            return false;
        }
    }
    return true;
}

static int testAllReduceFp16(ucclComm_t comm, sycl::queue& queue,
                              int rank, int nranks) {
    std::printf("[Rank %d] Testing AllReduce fp16...\n", rank);
    int errors = 0;

    /* Test various sizes */
    size_t testSizes[] = {1, 16, 256, 1024, 4096, 65536, 262144, 1048576};
    int nTests = sizeof(testSizes) / sizeof(testSizes[0]);

    for (int t = 0; t < nTests; t++) {
        size_t count = testSizes[t];

        /* Allocate device memory */
        sycl::half* sendbuff = sycl::malloc_device<sycl::half>(count, queue);
        sycl::half* recvbuff = sycl::malloc_device<sycl::half>(count, queue);

        /* Initialize: each rank fills with (rank + 1) */
        float initVal = static_cast<float>(rank + 1);
        queue.submit([&](sycl::handler& cgh) {
            cgh.parallel_for(sycl::range<1>(count),
                [=](sycl::id<1> i) {
                    sendbuff[i] = sycl::half(initVal);
                });
        }).wait();

        /* Run AllReduce */
        ucclResult_t res = ucclAllReduce(
            sendbuff, recvbuff, count,
            ucclFloat16, ucclSum, comm,
            static_cast<void*>(&queue));

        if (res != ucclSuccess) {
            std::fprintf(stderr,
                "[Rank %d] AllReduce failed for count=%zu: %s\n",
                rank, count, ucclGetErrorString(res));
            errors++;
        } else {
            queue.wait();

            /* Verify: sum of (rank+1) for all ranks
             * = 1 + 2 + ... + nranks = nranks*(nranks+1)/2 */
            float expectedSum = static_cast<float>(
                nranks * (nranks + 1) / 2);

            /* Copy result to host for verification */
            std::vector<sycl::half> hostResult(count);
            queue.memcpy(hostResult.data(), recvbuff,
                        count * sizeof(sycl::half)).wait();

            if (!checkResult(hostResult.data(), count,
                            expectedSum, 0.1f)) {
                std::fprintf(stderr,
                    "[Rank %d] FAIL: count=%zu, expected=%f\n",
                    rank, count, expectedSum);
                errors++;
            } else {
                std::printf("[Rank %d] PASS: count=%zu\n", rank, count);
            }
        }

        sycl::free(sendbuff, queue);
        sycl::free(recvbuff, queue);
    }

    return errors;
}

static int testAllReduceBf16(ucclComm_t comm, sycl::queue& queue,
                              int rank, int nranks) {
    std::printf("[Rank %d] Testing AllReduce bf16...\n", rank);
    int errors = 0;

    using bf16 = sycl::ext::oneapi::bfloat16;

    size_t count = 1024;

    bf16* sendbuff = sycl::malloc_device<bf16>(count, queue);
    bf16* recvbuff = sycl::malloc_device<bf16>(count, queue);

    float initVal = static_cast<float>(rank + 1);
    queue.submit([&](sycl::handler& cgh) {
        cgh.parallel_for(sycl::range<1>(count),
            [=](sycl::id<1> i) {
                sendbuff[i] = bf16(initVal);
            });
    }).wait();

    ucclResult_t res = ucclAllReduce(
        sendbuff, recvbuff, count,
        ucclBfloat16, ucclSum, comm,
        static_cast<void*>(&queue));

    if (res != ucclSuccess) {
        std::fprintf(stderr,
            "[Rank %d] AllReduce bf16 failed: %s\n",
            rank, ucclGetErrorString(res));
        errors++;
    } else {
        queue.wait();
        std::printf("[Rank %d] PASS: AllReduce bf16 count=%zu\n",
                    rank, count);
    }

    sycl::free(sendbuff, queue);
    sycl::free(recvbuff, queue);

    return errors;
}

int main(int argc, char** argv) {
    /* Only initialize MPI when this test is launched as multi-rank.
     * In single-process mode, avoid MPI+SYCL runtime interaction issues. */
    int rank = 0, nranks = 1;
    bool useMpi = false;
    bool mpiInitByUs = false;
    int mpiInit = 0;
    MPI_Initialized(&mpiInit);
    if (mpiInit) {
        useMpi = true;
    } else {
        const char* pmiSize = std::getenv("PMI_SIZE");
        const char* ompiSize = std::getenv("OMPI_COMM_WORLD_SIZE");
        int launcherSize = 1;
        if (pmiSize) launcherSize = std::atoi(pmiSize);
        else if (ompiSize) launcherSize = std::atoi(ompiSize);
        if (launcherSize > 1) {
            MPI_Init(&argc, &argv);
            useMpi = true;
            mpiInitByUs = true;
        }
    }

    if (useMpi) {
        MPI_Comm_rank(MPI_COMM_WORLD, &rank);
        MPI_Comm_size(MPI_COMM_WORLD, &nranks);
    }

    std::printf("[Rank %d/%d] Starting AllReduce test\n", rank, nranks);

    /* Select GPU */
    auto devices = sycl::device::get_devices(sycl::info::device_type::gpu);
    if (devices.empty()) {
        std::fprintf(stderr, "[Rank %d] No GPU devices found\n", rank);
        if (mpiInitByUs) MPI_Finalize();
        return 1;
    }

    int devIdx = rank % static_cast<int>(devices.size());
    sycl::queue queue(devices[devIdx]);

    /* Initialize UCCL communicator */
    ucclUniqueId id;
    if (rank == 0) ucclGetUniqueId(&id);
    if (useMpi) {
        MPI_Bcast(&id, sizeof(ucclUniqueId), MPI_BYTE, 0, MPI_COMM_WORLD);
    }

    ucclComm_t comm;
    ucclResult_t res = ucclCommInitRank(&comm, nranks, id, rank);
    if (res != ucclSuccess) {
        std::fprintf(stderr, "[Rank %d] CommInitRank failed: %s\n",
                     rank, ucclGetErrorString(res));
        if (mpiInitByUs) MPI_Finalize();
        return 1;
    }

    /* Run tests */
    int totalErrors = 0;
    totalErrors += testAllReduceFp16(comm, queue, rank, nranks);
    totalErrors += testAllReduceBf16(comm, queue, rank, nranks);

    /* Cleanup */
    ucclCommFinalize(comm);
    ucclCommDestroy(comm);

    /* Summary */
    int globalErrors = 0;
    if (useMpi) {
        MPI_Allreduce(&totalErrors, &globalErrors, 1,
                      MPI_INT, MPI_SUM, MPI_COMM_WORLD);
    } else {
        globalErrors = totalErrors;
    }

    if (rank == 0) {
        if (globalErrors == 0) {
            std::printf("\n=== ALL TESTS PASSED ===\n");
        } else {
            std::printf("\n=== %d TESTS FAILED ===\n", globalErrors);
        }
    }

    if (mpiInitByUs) MPI_Finalize();
    return (globalErrors == 0) ? 0 : 1;
}
